cpp/src/arrow/type.cc

```diff
@@ -42,6 +42,21 @@ std::shared_ptr<Field> Field::RemoveMetadata() const {
   return std::make_shared<Field>(name_, type_, nullable_);
 }
 
+std::vector<std::shared_ptr<Field>> Field::Flatten() const {
+  std::vector<std::shared_ptr<Field>> flattened;
+  if (type_->id() == Type::STRUCT) {
+    for (const auto& child : type_->children()) {
+      auto flattened_child = std::make_shared<Field>(*child);
+      flattened.push_back(flattened_child);
+      flattened_child->name_.insert(0, name() + ".");
```

Comment:
This is going to appear strange IMO if the name has a period or space or something other than variable name. This is kind of a can of worms without [`std::quoted`](http://en.cppreference.com/w/cpp/io/manip/quoted) (available in >=C++14)

Reply:
I'm not sure we should do anything about it. Names are indicative in any case, there can even be duplicate names ;-)